engine/gamesys/src/gamesys/resources/res_sound_data.cpp

```diff
@@ -41,6 +98,204 @@ namespace dmGameSystem
+        //DebugChunks(resource);
+        resource->m_RequestInFlight = 0;
+        return 1;
+    }
+
+    // Find the first chunk that contains the offset
+    static SoundDataChunk* FindChunk(SoundDataResource* resource, uint32_t offset)
+    {
+        uint32_t num_chunks = resource->m_Chunks.Size();
+        for (uint32_t i = 0; i < num_chunks; ++i)
+        {
+            SoundDataChunk* chunk = resource->m_Chunks[i];
+
+            if (offset < chunk->m_Offset)
+            {
+                // We currently don't have the chunk in question, and need to request it
+                return 0;
+            }
+            else if (offset < (chunk->m_Offset + chunk->m_Size))
+            {
+                return chunk;
+            }
+        }
+        return 0;
+    }
+
+    // Called from the Sound thread
+    static dmSound::Result SoundDataReadCallback(void* context, uint32_t offset, uint32_t size, void* _out, uint32_t* out_size)
+    {
+        SoundDataResource* resource = (SoundDataResource*)context;
+        DM_MUTEX_SCOPED_LOCK(resource->m_Context->m_Mutex);
+
+        if (offset >= resource->m_FileSize)
+        {
+            return dmSound::RESULT_END_OF_STREAM;
+        }
+
+        uint8_t* out = (uint8_t*)_out;
+
+        uint32_t nread = 0;
+        uint32_t bytes_to_read = size;
+
+        uint32_t request_size   = g_SoundDataContext->m_ChunkSize;
+        uint32_t request_offset = uint32_t(offset / request_size) * request_size;
+
+        SoundDataChunk* chunk = FindChunk(resource, offset);
+        while (chunk)
+        {
+            uint32_t chunk_size = chunk->m_Size;
+            uint32_t chunk_offset = chunk->m_Offset;
+
+            // We found a chunk and can start copying
+            uint32_t chunk_internal_offset = offset - chunk_offset; // the index to start from within this chunk
+            uint32_t chunk_remaining = chunk_size - chunk_internal_offset;
+            uint32_t to_read = dmMath::Min(bytes_to_read, chunk_remaining);
+            memcpy(out + nread, &chunk->m_Data[chunk_internal_offset], to_read);
+
+            bytes_to_read   -= to_read;
+            nread           += to_read;
+            offset          += to_read;
+
+            if (bytes_to_read > 0)
+                chunk = FindChunk(resource, offset);
+            else
+                chunk = 0;
+        }
+
+        uint32_t next_chunk_offset = uint32_t(offset / request_size) * request_size;
+        if (next_chunk_offset == request_offset)
+            next_chunk_offset += request_size;
+
+        if (next_chunk_offset >= resource->m_FileSize) // Are we wrapping around?
+        {
+            next_chunk_offset = 0;
+            request_offset = 0;
+        }
+
+        SoundDataChunk* next_chunk = FindChunk(resource, next_chunk_offset);
+        if (next_chunk && next_chunk->m_Offset != next_chunk_offset)
+        {
+            next_chunk = 0; // we're actually missing the next chunk. Let's trigger a pre fetch
+        }
```

Comment:
We may find a chunk as they're just stored sequentially in an array.
There is no guarantee it's the next chunk we want (as it may have been pruned earlier)